local.go

```diff
@@ -320,44 +339,92 @@ func evaluate(funcInfo *functionInfo, evalType EvaluationType, xNext []float64,
 	case GradEvaluation:
 		if funcInfo.IsGradient {
 			if different {
-				// Invalidate the function value because it will not be
-				// evaluated.
-				loc.F = math.NaN()
+				invalidate(loc, true, false, true)
 			}
 			funcInfo.gradient.Grad(loc.X, loc.Gradient)
 			stats.GradEvaluations++
 			return
 		}
 		if funcInfo.IsFunctionGradient {
+			if different {
+				invalidate(loc, false, false, true)
+			}
 			loc.F = funcInfo.functionGradient.FuncGrad(loc.X, loc.Gradient)
 			stats.FuncGradEvaluations++
 			return
 		}
+		if funcInfo.IsFunctionGradientHessian {
+			if loc.Hessian == nil {
```

Comment:
I believe this is here, because the Hessian is only allocated if the method needs a hessian. Given this change, shouldn't IsFunctionGradient allocate a gradient if it's nil?